### PR TITLE
refactor: split paren parser

### PR DESCRIPTION
**Pull request by Kha**
Experiment based on #1799

### CODE REVIEW

**Comment by Kha** [comment]
!bench

**Comment by Kha** [comment]
Wait... it got faster?

File: src/Lean/Parser/Term.lean [hunk 1]
```diff
@@ -137,8 +135,8 @@ Parentheses, used for
   - `(f · a b)` is shorthand for `fun x => f x a b`
   - `(h (· + 1) ·)` is shorthand for `fun x => h (fun y => y + 1) x`
 -/
-@[builtin_term_parser] def paren := leading_parser
-  "(" >> (withoutPosition (withoutForbidden (optional (ppDedentIfGrouped termParser >> parenSpecial)))) >> ")"
+@[builtin_term_parser] def paren := (leading_parser
```
Comment (gebner) [inline src/Lean/Parser/Term.lean]: If we're already splitting this parser, we should actually split it in three declarations.

Comment (gebner) [reply src/Lean/Parser/Term.lean]: Nevermind, I didn't see part 2.

Comment (Kha) [reply src/Lean/Parser/Term.lean]: Ah, there is a fun development story hidden here: given that the paren-like syntaxes are used in quotations all over the place, it was very important to make them use the new parser from the same stage that their elaborators were adapted in. But `parseQuotWithCurrentStage` does not affect *newly* added category parsers, thus the intermediate solution above. All in all, the code manually building syntax trees of the affected kinds was definitely the bigger issue as you have to locate it in the first place before you can fix it.

**Pull request merged**